sky/clouds/kubernetes.py

```diff
@@ -169,6 +180,36 @@ def existing_allowed_contexts(cls) -> List[str]:
         cls._log_skipped_contexts_once(tuple(skipped_contexts))
         return existing_contexts
 
+    @classmethod
+    def _log_unavailable_context(cls,
+                                 context: str,
+                                 reason: Optional[str] = None) -> None:
+        """Logs a Kubernetes context as unavailable.
+
+        Args:
+            context: The Kubernetes context to mark as unavailable.
+            reason: Optional reason for marking the context as unavailable.
+            silent: Whether to suppress the log message.
+        """
+        # Skip if this context has already been logged as unavailable
+        if context in cls.logged_unavailable_contexts:
+            return
+
+        cls.logged_unavailable_contexts.add(context)
+        msg = f'Excluding Kubernetes context {context}'
+        if reason is not None:
+            msg += f': {reason}'
+        logger.info(msg)
+
+        # Check if all existing allowed contexts are now unavailable
+        existing_contexts = cls.existing_allowed_contexts()
+        if existing_contexts and all(ctx in cls.logged_unavailable_contexts
+                                     for ctx in existing_contexts):
+            logger.warning(
+                'All Kubernetes contexts have reported as unavailable. '
+                'Retry if it is a transient error, or run sky check to '
```

Comment:
Is it possible that this is not only a transient error but also a normal resource unavailable error, i.e. all GPUs are occupied at the moment, considering we are catching the entire `ResourcesUnavailableError`?

Reply:
The short answer is no.

This reminds me of our previous discussion https://github.com/skypilot-org/skypilot/pull/4692#discussion_r1951866258

> - The exception will be raised when there is an error contacting the k8s API
> - fits=False will be returned if there is no fit, no exception will be raised

The handling here depends on this global knowledge, which is weird indeed. Just refined the handling